File: python/spider_browser/ai/llm_provider.py
"""Pluggable LLM provider interface."""

from __future__ import annotations

import json
import re
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, Dict, List, Literal, Optional, Union

from ..utils.errors import LLMError


@dataclass
class LLMConfig:
    """LLM configuration for AI methods."""

    provider: Literal["openai", "anthropic", "openrouter"]
    model: str
    api_key: str
    base_url: Optional[str] = None
    max_tokens: int = 4096
    temperature: float = 0.1


# Message types matching OpenAI's format
LLMContentPart = Union[
    Dict[str, Any],  # {"type": "text", "text": "..."} or {"type": "image_url", "image_url": {"url": "..."}}
]

LLMMessage = Dict[str, Any]  # {"role": "system"|"user"|"assistant", "content": str | list[LLMContentPart]}


class LLMProvider(ABC):
    """Abstract LLM provider."""

    @abstractmethod
    async def chat(self, messages: List[LLMMessage], *, json_mode: bool = False) -> str:
        """Call the LLM with messages and get a text response."""
        ...

    async def chat_json(self, messages: List[LLMMessage]) -> Any:
        """Call the LLM and parse the response as JSON."""
        text = await self.chat(messages, json_mode=True)
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            # Try extracting from markdown code blocks
            match = re.search(r"```(?:json)?\s*([\s\S]*?)```", text)
            if match:
                return json.loads(match.group(1))
            raise LLMError(f"LLM response is not valid JSON: {text[:200]}")


def create_provider(config: LLMConfig) -> LLMProvider:
    """Create an LLM provider from config."""
    if config.provider in ("openai", "openrouter"):
        from .providers.openai_provider import OpenAICompatibleProvider
        return OpenAICompatibleProvider(config)
    elif config.provider == "anthropic":
        from .providers.anthropic_provider import AnthropicProvider
        return AnthropicProvider(config)
    else:
        raise ValueError(f"Unknown LLM provider: {config.provider}")

File: go/llm_provider.go
package spiderbrowser

// LLMConfig configures an LLM provider for AI methods.
type LLMConfig struct {
	// Provider: "openai", "anthropic", or "openrouter".
	Provider string
	// Model name (e.g. "gpt-4o", "claude-sonnet-4-5-20250929").
	Model string
	// APIKey for the provider.
	APIKey string
	// BaseURL override (e.g. for OpenRouter or local vLLM).
	BaseURL string
	// MaxTokens (default: 4096).
	MaxTokens int
	// Temperature (default: 0.1).
	Temperature float64
}

func (c *LLMConfig) defaults() {
	if c.MaxTokens == 0 {
		c.MaxTokens = 4096
	}
	if c.Temperature == 0 {
		c.Temperature = 0.1
	}
}

// LLMMessage is a message for LLM calls.
type LLMMessage struct {
	Role    string       `json:"role"` // "system", "user", "assistant"
	Content any          `json:"content"` // string or []LLMContentPart
}

// LLMContentPart is a part of a multimodal message.
type LLMContentPart struct {
	Type     string        `json:"type"` // "text" or "image_url"
	Text     string        `json:"text,omitempty"`
	ImageURL *LLMImageURL  `json:"image_url,omitempty"`
}

// LLMImageURL is an image URL in a content part.
type LLMImageURL struct {
	URL string `json:"url"`
}

// LLMProvider is the interface for LLM providers.
type LLMProvider interface {
	// Chat calls the LLM with messages and returns a text response.
	Chat(messages []LLMMessage, jsonMode bool) (string, error)
	// ChatJSON calls the LLM and parses the response as JSON into result.
	ChatJSON(messages []LLMMessage, result any) error
}

// CreateProvider creates an LLM provider from config.
func CreateProvider(config LLMConfig) LLMProvider {
	config.defaults()
	switch config.Provider {
	case "openai", "openrouter":
		return NewOpenAIProvider(config)
	case "anthropic":
		return NewAnthropicProvider(config)
	default:
		panic("unknown LLM provider: " + config.Provider)
	}
}
